### DISCUSSION TITLE
Cannot remove plugins from Theia IDE

### DISCUSSION THREAD

**Discussion by dagoston93** [eclipse-theia/theia#13690] [General]
Hi,

I created an IDE project by downloading the Theia IDE repo and customize it as necessary.
I managed to add some custom features, change the branding, but I have problems with removing stuff...

First I tried to remove some Themes (for example Abyss, Red, Solarized, etc...), but failed to do so.
First I tried to add the package ids to the `theiaPluginsExcludeIds` properties of the main **package.json** file. But after running **yarn clean**, deleting the **plugins** folder, and then re-building the themes still exists in the **File/Preferences/Color Themes**
Then I tried to remove the whole `eclipse-theia.builtin-extension-pack` that includes them, thinking that I will re-add plugins I need, but after re-building the same way, no effect...

Then I thought I would try to remove a couple of views that I do not need. But again, failure...
What I did was, to remove some dependencies from the **applications/electron/package.json** file.
For example:

- @theia/callhierarchy
- @theia/outline-view
- and a couple of others...

But again after rebuilding, nothing happens, the views are still available, and even more strangely, the plugins are downloaded to the **plugins** folder...

What am I doing wrong, and how to remove the items I do not need?

Thanks in advance for any help,
Best regards,
Agoston

**Comment by msujew** [comment]
Hey @dagoston93,

> Then I thought I would try to remove a couple of views that I do not need. But again, failure...

Since all of these packages are required during compile time by the `@theia/plugin-ext` package (that provides the plugin support), removing them doesn't do anything, as they will be transitively pulled anyway.

We have a [contribution filter API](https://github.com/eclipse-theia/theia/blob/master/examples/api-samples/src/browser/contribution-filter/sample-filtered-command-contribution.ts) to prevent certain widgets/commands/menus from being contributed. You should use that instead.

> But again after rebuilding, nothing happens, the views are still available, and even more strangely, the plugins are downloaded to the plugins folder...

What kind of IDs are you using for the `theiaPluginsExcludeIds` property? They need the full id, i.e. `<publisher>.<name>`.

In case that doesn't help, please provide a reproducible example repo.

**Reply by dagoston93** [reply]
Hi, 
thanks for the quick reply.

I used the following syntax:
```
  "theiaPluginsExcludeIds": [
        "vscode.theme-abyss",
        "vscode.theme-monokai"
   ]
```
But it did nothing... So do I need to remove the themes also using the contributin filter API?

Do I really need the @theia/plugin-ext? Because I generated a theia extension with the **yo** generator and it generated it, and it doesn't include it, but the application that it also generated works fine... So what kind of plugin support does it exactly do? The support of VS Code plugins?

Thanks in advance,
BR,
Agoston

**Reply by msujew** [reply]
> But it did nothing...

Note that it doesn't delete plugins if they already exist on disk. You will need to delete them manually from disk.

> So do I need to remove the themes also using the contributin filter API?

No, the contribution filter API is only for compile-time extensions. Plugins are runtime contributed.

> So what kind of plugin support does it exactly do? The support of VS Code plugins?

Yes, without it, VSCode extensions aren't loaded.

**Reply by dagoston93** [reply]
Hi,
Finally back at the project, I had some different topics to work on in the past days...
So I managed to figure out, that somehow the themes and probably some other plugins get cached, or I don't know, because if I do a complete re-build (delete node_modules, package, dist folders, then build) and then I start the app with the **yarn electron start** command, the themes are still there, but if I run the installer, then they are excluded in the installed IDE, as I want them to be...
Why is it so, and how can I prevent it to be the case?
It is a bit cumbersome to uninstall/install my IDE after each change...

**Reply by msujew** [reply]
@dagoston93 Where does your [`theiaPluginsDir`](https://github.com/eclipse-theia/theia/blob/dacc1aeda03290ce49a3cd02af180ee20c2a8092/package.json#L103) point to? That's the directory where Theia stores its builtin plugins.

**Reply by dagoston93** [reply]
It is left as defaule: "theiaPluginsDir": "plugins".
But if it refers to the plugins dir of the repo's root, I always delete it... Or maybe **yarn electron start** copies it somewhere?

**Reply by msujew** [reply]
> Or maybe yarn electron start copies it somewhere?

It doesn't. Deleting the `plugins` directory should have the desired effect. I've just confirmed that deleting the `plugins` directory correctly disables all builtin extensions. It might be that you have a duplicate of the extension in your `~/.theia` directory, where Theia stores user installed extensions. Can you delete that to see whether that fixes the issue?

**Reply by dagoston93** [reply]
Interesting... I only have a .theia folder in my user folder (I'm on Windows 10), but if I empty it, that doesn't solve it... It only seem to work with the installed version... :( (also ,for that I need to delete the folder it creates in AppData/Roaming...)

**Reply by msujew** [reply]
When you open the `Extensions` view in the side panel, do you have any extensions visible after deleting the `plugins` folder?

**Reply by dagoston93** [reply]
Yes, the panel works fine, it shows up results if I search for plugins. (No extensions are installed, and none were installed before.)

**Reply by msujew** [reply]
The panel also shows builtin extensions. Are there any shown even though you've deleted the `plugins` folder? That's the behavior you're experiencing right? Builtins appearing in the app, but not visible on the file system.

**Reply by dagoston93** [reply]
Ohh, I found the built-in tab...
So if I delete the dist/win-unpacked/resources/app/plugins folder, then no built-in extensions show up, but the themes are still there...
I don't get it how it works... I mean in theory I guess I understand, but theory doesn't seem to match reality...

Also I am trying to use FilterContribution API, but struggling there too...
I created a filter with the following content:
```
import { ContributionFilterRegistry, FilterContribution } from "@theia/core";
import { injectable } from "inversify";
import { OutlineViewService } from "@theia/outline-view/src/browser/outline-view-service";
import { OutlineViewContribution } from "@theia/outline-view/src/browser/outline-view-contribution";
import { OutlineBreadcrumbsContribution } from "@theia/outline-view/src/browser/outline-breadcrumbs-contribution";

@injectable()
export class SwcTstContributionFilter implements FilterContribution {
    registerContributionFilters(registry: ContributionFilterRegistry): void {
        // Remove Outline view
        registry.addFilters("*", [
            (contrib) => contrib instanceof OutlineViewService,
            (contrib) => contrib instanceof OutlineViewContribution,
            (contrib) => contrib instanceof OutlineBreadcrumbsContribution,
        ]);
    }
}
```

But when I try to build, I get an error: **Can't resolve '@theia/outline-view/src/browser/outline-view-service' in 'C:\Company\projects\TestFramework\Frontend\SwcTestFrameworkUI\theia-extensions\product\lib\browser'**

I guess I understand why it is... The contributions I am trying to filter are in the node_modules folder of the main project, and this file I created is in the theia_extensions/product folder, thus, I guess it accesses only the node_modules folder in the extension's folder...
But I have no clue how to overcome this... Is this even the right way to use the filter API to filter out unwanted contris?

BR, Agoston

**Reply by msujew** [reply]
> but the themes are still there

Which themes are still there? Theia comes with 3 base themes (dark/light/high contrast)

> But when I try to build, I get an error: Can't resolve '@theia/outline-view/src/browser/outline-view-service' in 'C:\Company\projects\TestFramework\Frontend\SwcTestFrameworkUI\theia-extensions\product\lib\browser'

You need to import from `@theia/outline-view/lib/browser/outline-view-service`, as the `src` folder does not contain JavaScript code.

**Reply by dagoston93** [reply]
> Which themes are still there? Theia comes with 3 base themes (dark/light/high contrast)

Abyss, Solarized Light, Solarized Dark, Kimbie... such themes...

> You need to import from @theia/outline-view/lib/browser/outline-view-service, as the src folder does not contain JavaScript code.

Thanks, that way it compiles, but as the program loads, I only get a white screen and some errors in the console's log, from which I call **yarn electron start**.
```
2024-05-16T15:05:00.721Z root WARN Couldn't restore widget for test-view-container. Error: Error: No widget factory 'test-view-container' has been registered.
2024-05-16T15:05:00.722Z root WARN Couldn't restore widget for vsx-extensions-view-container. Error: Error: No widget factory 'vsx-extensions-view-container' has been registered.
2024-05-16T15:05:00.722Z root WARN Couldn't restore widget for debug. Error: Error: No widget factory 'debug' has been registered.
2024-05-16T15:05:00.723Z root WARN Couldn't restore widget for scm-view-container. Error: Error: No widget factory 'scm-view-container' has been registered.
2024-05-16T15:05:00.723Z root WARN Couldn't restore widget for search-view-container. Error: Error: No widget factory 'search-view-container' has been registered.
2024-05-16T15:05:00.723Z root WARN Couldn't restore widget for explorer-view-container. Error: Error: No widget factory 'explorer-view-container' has been registered.
2024-05-16T15:05:00.723Z root WARN Couldn't restore widget for problems. Error: Error: No widget factory 'problems' has been registered.
2024-05-16T15:05:00.724Z root WARN Couldn't restore widget for getting.started.widget. Error: Error: No widget factory 'getting.started.widget' has been registered. 
```

But I did not even touch those widgets...
(I also added the following filter: `(contrib) => contrib instanceof OutlineViewWidget,` .

**Reply by msujew** [reply]
> Abyss, Solarized Light, Solarized Dark, Kimbie... such themes...

I see. Without literally looking into your file system, I have no idea where these come from. Do they really not appear in the `builtins` of the extensions view?

> Thanks, that way it compiles, but as the program loads, I only get a white screen and some errors in the console's log, from which I call yarn electron start.

cc @tortmayr any idea? I'm not too familiar with the contribution filter API.

**Reply by tortmayr** [reply]
@dagoston93 You need to negate your filter predicates. The contribution filter function behaves like the Array.filter function i.e. it includes everything that evaluates to `true` and filters everything that evaluates to `false`.
In addition, I also recommend to filter out the `WidgetFactory` of the outline view. Otherwise you might run into the issue that the outlineview is restored from the local storage.

Something like this should work:
```ts

import { ContributionFilterRegistry, FilterContribution } from "@theia/core";
import { injectable } from "inversify";
import { OutlineViewService } from "@theia/outline-view/lib/browser/outline-view-service";
import { OUTLINE_WIDGET_FACTORY_ID, OutlineViewContribution } from "@theia/outline-view/lib/browser/outline-view-contribution";
import { OutlineBreadcrumbsContribution } from "@theia/outline-view/lib/browser/outline-breadcrumbs-contribution";
import { WidgetFactory } from '@theia/core/lib/browser';

@injectable()
export class SwcTstContributionFilter implements FilterContribution {
    registerContributionFilters(registry: ContributionFilterRegistry): void {
        // Remove Outline view
        registry.addFilters("*", [
            (contrib) => !(contrib instanceof OutlineViewService),
            (contrib) => !(contrib instanceof OutlineViewContribution),
            (contrib) => !(contrib instanceof OutlineBreadcrumbsContribution),
        ]);
        registry.addFilters(
            [WidgetFactory],
            [filterById(id => OUTLINE_WIDGET_FACTORY_ID !== id)]
        );

    }
}

function filterById(filter: Filter<string>): Filter<object & { id?: string }> {
    return object =>
        object.id ? filter(object.id) : true;

}

````

**Reply by dagoston93** [reply]
> You need to negate your filter predicates...

Thank you, that solves it :)

> Do they really not appear in the builtins of the extensions view?

If I delete the plugins folder, then they are not there, but show up in the Preferences/Color Theme...
Anyway, I will try to clean my file system, because, I had installation of Theia IDE, the customized IDE and it's git repo... Somewhere, something must be stuck...

Thank you both for the help :)

**Reply by dagoston93** [reply]
And one more thing came up?
How do I remove preferences?
I managed to remove the **@theia/debug** module using the above technique, but it has a `bindDebugPreferences()` method, and uses the `bind(DebugPreferences).toDynamicValue()` to bind the preferences, but I cannot use that technique to filter out `DebugPreferences` because I get a compile error.
How can I get around this?

**Reply by msujew** [reply]
I wouldn't recommend removing preferences as most service implementations are written with preferences having default values in mind. I.e. you might accidentally breaking features with that. Instead, I would recommend hiding them. For example for some of the debug preferences:

```ts
import { debugPreferencesSchema } from '@theia/debug/lib/browser/debug-preferences';
debugPreferencesSchema.properties['debug.trace'].included = false;
```

That way they are hidden from the user but still available programmatically with their default values.

**Reply by dagoston93** [reply]
Thanks, that works fine.

One more problem I encounter...
I am trying to exclude **ms-vscode.js-debug** and **ms-vscode.js-debug-companion** plugins, I add them to the **theiaPluginsExcludeIds** section of **package.json**, but unlike other extensions, they still get downloaded...

How can I prevent it?
Are they special in any way?
I did search for their package name, but the only reference to it are in **eclipse-theia.builtin-extension-pack** (which shall not cause problems, as I exclude successfully other packages referenced by it), and in my **package.json** where I exclude it...

Thanks in advance for your answer,
BR,
Agoston
